### PR TITLE
Adding mutual tls with ID Secret authenticator

### PR DESCRIPTION
**Pull request by nuwandiw**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/pom.xml [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+            </plugin>
+            <plugin>
+                <groupId>org.apache.felix</groupId>
+                <artifactId>maven-bundle-plugin</artifactId>
+                <extensions>true</extensions>
+                <configuration>
+                    <instructions>
+                        <Bundle-SymbolicName>${project.artifactId}</Bundle-SymbolicName>
+                        <Bundle-Name>${project.artifactId}</Bundle-Name>
+                        <Private-Package>
```
Comment (hasinthaindrajee) [inline component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/pom.xml]: Why don't we have imports here ? Don't we use any osgi imports ?

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+import java.security.cert.X509Certificate;
+import java.util.List;
+import java.util.Map;
+
+import static org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret.util.MutualTLSUtil.JAVAX_SERVLET_REQUEST_CERTIFICATE;
+
+/**
+ * This class is an enhancement for BasicAuthClientAuthenticator which validates the client certificate
+ * in addition to the client id and secret authentication
+ *
```
Comment (hasinthaindrajee) [inline component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java]: Remove extra line

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+            X509Certificate registeredCert = null;
+            try {
+                registeredCert = (X509Certificate) OAuth2Util
+                        .getX509CertOfOAuthApp(oAuthClientAuthnContext.getClientId(), tenantDomain);
+            } catch (IdentityOAuth2Exception e) {
+                if (e.getCause() instanceof CertificateException) {
+                    throw e;
+                } else {
+                    // This means certificate is not configured in service provider. In that case basic authentication
+                    // would be performed
```
Comment (hasinthaindrajee) [inline component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java]: Can we add a debug log here to see what is actually taking place ? Also at debug level you may print the exception as well.

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+                    return super.authenticateClient(request, bodyParams, oAuthClientAuthnContext);
```
Comment (hasinthaindrajee) [inline component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java]: Do we need to re-authenticate here ?  We have already passed basic authentication

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/internal/MutualTLSServiceComponent.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    private static Log log = LogFactory.getLog(MutualTLSServiceComponent.class);
+    private BundleContext bundleContext;
+
+    @Activate
+    protected void activate(ComponentContext context) {
+
+        try {
+            // Registering MutualTLSClientAuthenticator as an OSGIService.
+            bundleContext = context.getBundleContext();
+            MutualTLSWithIdSecretAuthenticator mutualTLSWithIdSecretAuthenticator = new MutualTLSWithIdSecretAuthenticator();
```
Comment (hasinthaindrajee) [inline component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/internal/MutualTLSServiceComponent.java]: Doesn't this exceed the limit ?

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java [hunk 6]
```diff
@@ -0,0 +1,186 @@
...
+
+    /**
+     * Returns the execution order of this authenticator
+     *
+     * @return Execution place within the order
+     */
+    @Override
+    public int getPriority() {
+
+        return 101;
```
Comment (hasinthaindrajee) [inline component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java]: Any idea why we need to hard code this priority ?

Comment (nuwandiw) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java]: We have given hardcoded getPriority() value 100 for BasicAuthClientAuthenticator (Extended class). If this value is not provided, same priority would be given to git authenticator.

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java (hunk 3 above)
Comment (nuwandiw) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/src/main/java/org/wso2/carbon/identity/oauth2/token/handler/clientauth/tlswithidsecret/MutualTLSWithIdSecretAuthenticator.java]: Added

File: component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/pom.xml (hunk 1 above)
Comment (nuwandiw) [reply component/org.wso2.carbon.identity.oauth2.token.handler.clientauth.tlswithidsecret/pom.xml]: Added imports

**Pull request merged**